test/integration/auth/auth_test.go

```diff
@@ -875,6 +875,74 @@ func TestImpersonateIsForbidden(t *testing.T) {
 
 }
 
+func TestImpersonateWithUID(t *testing.T) {
+
+	masterConfig := framework.NewIntegrationTestMasterConfig()
+	masterConfig.GenericConfig.Authentication.Authenticator = getTestTokenAuth()
+	masterConfig.GenericConfig.Authorization.Authorizer = impersonateAuthorizer{}
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+	defer closeFn()
+
+	transport := http.DefaultTransport
+
+	var csrBody = `
+{
+  "kind": "CertificateSigningRequest",
+  "apiVersion": "certificates.k8s.io/v1",
+  "metadata": {
+    "name": "john"
+  },
+  "spec": {
+    "groups": ["system:authenticated"],
+    "signerName": "kubernetes.io/kube-apiserver-client",
+    "request": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURSBSRVFVRVNULS0tLS0KTUlJQ1ZqQ0NBVDRDQVFBd0VURVBNQTBHQTFVRUF3d0dZVzVuWld4aE1JSUJJakFOQmdrcWhraUc5dzBCQVFFRgpBQU9DQVE4QU1JSUJDZ0tDQVFFQTByczhJTHRHdTYxakx2dHhWTTJSVlRWMDNHWlJTWWw0dWluVWo4RElaWjBOCnR2MUZtRVFSd3VoaUZsOFEzcWl0Qm0wMUFSMkNJVXBGd2ZzSjZ4MXF3ckJzVkhZbGlBNVhwRVpZM3ExcGswSDQKM3Z3aGJlK1o2MVNrVHF5SVBYUUwrTWM5T1Nsbm0xb0R2N0NtSkZNMUlMRVI3QTVGZnZKOEdFRjJ6dHBoaUlFMwpub1dtdHNZb3JuT2wzc2lHQ2ZGZzR4Zmd4eW8ybmlneFNVekl1bXNnVm9PM2ttT0x1RVF6cXpkakJ3TFJXbWlECklmMXBMWnoyalVnald4UkhCM1gyWnVVV1d1T09PZnpXM01LaE8ybHEvZi9DdS8wYk83c0x0MCt3U2ZMSU91TFcKcW90blZtRmxMMytqTy82WDNDKzBERHk5aUtwbXJjVDBnWGZLemE1dHJRSURBUUFCb0FBd0RRWUpLb1pJaHZjTgpBUUVMQlFBRGdnRUJBR05WdmVIOGR4ZzNvK21VeVRkbmFjVmQ1N24zSkExdnZEU1JWREkyQTZ1eXN3ZFp1L1BVCkkwZXpZWFV0RVNnSk1IRmQycVVNMjNuNVJsSXJ3R0xuUXFISUh5VStWWHhsdnZsRnpNOVpEWllSTmU3QlJvYXgKQVlEdUI5STZXT3FYbkFvczFqRmxNUG5NbFpqdU5kSGxpT1BjTU1oNndLaTZzZFhpVStHYTJ2RUVLY01jSVUyRgpvU2djUWdMYTk0aEpacGk3ZnNMdm1OQUxoT045UHdNMGM1dVJVejV4T0dGMUtCbWRSeEgvbUNOS2JKYjFRQm1HCkkwYitEUEdaTktXTU0xMzhIQXdoV0tkNjVoVHdYOWl4V3ZHMkh4TG1WQzg0L1BHT0tWQW9FNkpsYWFHdTlQVmkKdjlOSjVaZlZrcXdCd0hKbzZXdk9xVlA3SVFjZmg3d0drWm89Ci0tLS0tRU5EIENFUlRJRklDQVRFIFJFUVVFU1QtLS0tLQo=",
+    "usages": ["client auth"]
+  }
+}
+`
+	csrCreateUrl := "/apis/certificates.k8s.io/v1/certificatesigningrequests"
+
+	token := BobToken
+	createCSRBodyBytes := bytes.NewReader([]byte(csrBody))
+	createCSRReq, err := http.NewRequest("POST", s.URL+csrCreateUrl, createCSRBodyBytes)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	createCSRReq.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+	createCSRReq.Header.Set("Impersonate-User", "alice")
+	createCSRReq.Header.Set("Impersonate-UID", "1")
```

Comment:
Use the standard typed CSR client with an impersonating config + build it from a rest config with a `.Wrap` func that sets the UID as well.